### PR TITLE
Migrate to asto 0.11.0

### LINKED ISSUE #14: Migrate to asto 0.11
In order to keep the repository consistent with latest asto changes, we should migrate to the latest version.

### PR DESCRIPTION
**Pull request by Sammers21**
fixes #14 

Changes done is this pull request: the version of asto has been changed from 0.2 to 0.11. Sources have been updated according to asto changes.

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#15](https://github.com/artipie/npm-adapter/issues/15) is assigned to @iakunin[/z](https://www.0crat.com/u/iakunin), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/cee3f8f3-4bfe-4df3-84b7-7b6dd663e249); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
@g4s8[/z](https://www.0crat.com/u/g4s8) I see that you unassigned this issue; the order is still assigned to @iakunin[/z](https://www.0crat.com/u/iakunin) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 11@952946a59154, thread: PQ-CT2E6TK9B
com.zerocracy.claims.ClaimIn[214] java.lang.IllegalArgumentException: Parameter "age" not found in "Order was canceled" among: [job, reason, cause, _parent_sig, flow, login]
```

[0.51.0](https://github.com/zerocracy/farm): CID: [109c3bd7-45ea-494a-831c-bad58b9e117a](https://www.0crat.com/footprint/CT2E6TK9B/109c3bd7-45ea-494a-831c-bad58b9e117a), Type: "Order was canceled"

**Comment by 0crat** [comment]
This pull request [#15](https://github.com/artipie/npm-adapter/issues/15) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/c2510965-bac9-4e89-ba71-544482b33ea4); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

File: .travis.yml [hunk 1]
```diff
@@ -11,14 +13,13 @@ cache:
...
   - rvm use 2.6.0
-  - sudo apt-get -y install nodejs-dev node-gyp libssl1.0-dev openjdk-8-jdk
+  - sudo apt-get -y install nodejs-dev node-gyp libssl1.0-dev
   - sudo apt-get -y install npm
 install:
   - gem install pdd -v 0.20.5
   - gem install xcop
 script:
   - set -e
-  - JAVA_HOME="/usr/lib/jvm/java-8-openjdk-amd64"
```
Comment (amihaiemil) [inline .travis.yml]: @Sammers21 You replaced jdk8 with jdk9, but why did you remove JAVA_HOME?

File: pom.xml [hunk 2]
```diff
@@ -97,14 +98,24 @@ SOFTWARE.
...
-      <groupId>com.yegor256</groupId>
+      <groupId>com.artipie</groupId>
       <artifactId>asto</artifactId>
-      <version>0.2.0</version>
+      <version>0.11.0</version>
+    </dependency>
+    <dependency>
+      <groupId>com.github.akarnokd</groupId>
+      <artifactId>rxjava3-jdk8-interop</artifactId>
+      <version>${rxjava.version}</version>
```
Comment (amihaiemil) [inline pom.xml]: @Sammers21 What's this library for?

File: src/main/java/com/artipie/npm/Meta.java [hunk 3]
```diff
@@ -42,14 +41,14 @@
     /**
      * The meta.json file.
      */
-    private final Path json;
+    private final JsonObject json;
```
Comment (amihaiemil) [inline src/main/java/com/artipie/npm/Meta.java]: @Sammers21 What is this after all? Is it a Path or a JsonObject...? Why did you change?

File: src/main/java/com/artipie/npm/Meta.java [hunk 4]
```diff
@@ -60,29 +59,32 @@
...
-                Files.write(
-                    this.json,
-                    patch
-                        .build()
-                        .apply(meta)
-                        .toString()
-                        .getBytes(StandardCharsets.UTF_8)
-                );
-            });
+    public Meta updatedMete(final JsonObject uploaded) {
```
Comment (amihaiemil) [inline src/main/java/com/artipie/npm/Meta.java]: @Sammers21 Why is this method name in past tence? I think it should be called ``updateMeta`` and the json argument should be called ``patch`` (it is a patch which is being added to the current meta, right?)

File: pom.xml (hunk 2 above)
Comment (Sammers21) [reply pom.xml]: A dependency from asto. Forced to be declared by qulice

File: .travis.yml (hunk 1 above)
Comment (Sammers21) [reply .travis.yml]: Bionic distribution did not have support for java 8, that is why it was installed manually.

File: src/main/java/com/artipie/npm/Meta.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/com/artipie/npm/Meta.java]: It is a JsonObject. 

> Why did you change

Using Path was forced by the previous version of asto. Files were used, now we can store it in memory

File: src/main/java/com/artipie/npm/Meta.java (hunk 4 above)
Comment (Sammers21) [reply src/main/java/com/artipie/npm/Meta.java]: Disagree. `update mete` is a verb + noun, `updated meta` is  an adjective + noun. According to EO, manipulator methods should return void.

Comment (Sammers21) [reply src/main/java/com/artipie/npm/Meta.java]: > json argument should be called patch

Disagree. `uploaded` is more specific

Comment (amihaiemil) [reply src/main/java/com/artipie/npm/Meta.java]: @Sammers21 So you want to read this method as "get me the updated meta" -- fair. But still I think the argument should be called ``patch``? -- as in "Get me the updated meta with this patch". Or?

**Comment by amihaiemil** [comment]
@Sammers21 see my comment... also take care of the descriptions, maybe add some unit tests, or make more smaller PRs. This here is huge, and I don't understand why so many code changes are necessary for updating a dependency.

**Comment by Sammers21** [comment]
@amihaiemil,  PR and issue description have been updated.

File: src/main/java/com/artipie/npm/Meta.java (hunk 4 above)
Comment (Sammers21) [reply src/main/java/com/artipie/npm/Meta.java]: `uploaded` is more specific, because this exact JSON has been uploaded by the `npm install` command. If we would name it `a patch`, it would be confusing to read.

**Comment by Sammers21** [comment]
> maybe add some unit tests,

What are the reason to do this? There is no new functionality, introduced in this Pull Request. 

> or make more smaller PRs

No, it is already as small as possible. Most of the changes are indention level changes, so efficiency-wise, it is a small pull request.

**Comment by Sammers21** [comment]
> understand why so many code changes are necessary for updating a dependency

The old versions of asto were using files for storing artifacts. The newer versions are using Java 9 Flows for data streams. It is more flexible.

**Comment by Sammers21** [comment]
@amihaiemil , review the Pull Request again, please.

**Comment by amihaiemil** [comment]
@Sammers21 You probably did not add any new functionality, but you changed the implementation of existing functionalities! And as I see, there are no unit tests for the classes you have changed in this PR: https://github.com/artipie/npm-adapter/tree/master/src/test/java/com/artipie/npm

Please leave a TODO somewhere, for future testing: 

```
@todo #14:30min We need to unit test all our classes and methods, since we currently
 have almost no tests at all. Please also think of integration testing and adding a Coverage
 badge in the README. We can use Jacoco for test coverage calculations and Coveralls.io
 for displaying a badge. 
```

Copy the above TODO exactly as it is, including indentation and spacing, and add it in the code, in a JavaDoc somewhere. Then, when this PR is merged, a Github Issue will be opened automatically by 0pdd.

**Comment by Sammers21** [comment]
> And as I see, there are no unit tests for the classes you have changed in this PR:

There is a test: `NpmCommandsTest`, the test ensures that the existed implementation is compatible with `npm install` and `npm publish` command, which is the main requirement for the project.

**Comment by amihaiemil** [comment]
@Sammers21 Just leave the todo please, and I'll aprove the PR. We need tests. Having a few tests in that class doesn't cover everything, or if it does, it means they are Integration Tests. We also need Unit Tests.

**Comment by Sammers21** [comment]
> We need tests. Having a few tests in that class doesn't cover everything, or if it does, it means they are Integration Tests.

I am not convinced we need more tests since the main functionality is already covered. Setting up test coverage metrics would useful indeed. But it is not a goal of this pull request. The primary goal of this pull request is to migrate to the latest version of asto, not to add a TODO, not to discuss current test coverage or anything else.

**Comment by amihaiemil** [comment]
@Sammers21 You should add the TODO in this Pull Request because you did not write tests in this Pull Requests. Tests are expected with any code change. 

So because you already spent a lot of time with this PR, you leave a TODO in order for the work to be delegated in a future task!

**Comment by g4s8** [comment]
@Sammers21 I agree with @amihaiemil let's add a todo for test

**Comment by amihaiemil** [comment]
@g4s8 Why the hell do DEVs have the right to merge PRs? :)

**Comment by Sammers21** [comment]
>You should add the TODO in this Pull Request because you did not write tests in this Pull Requests. Tests are expected with any code change.
>
>So because you already spent a lot of time with this PR, you leave a TODO in order for the work to be delegated in a future task!
>
> @Sammers21 I agree with @amihaiemil let's add a todo for test

@amihaiemil , @g4s8 , you are welcome to create a new issue in this repository. As it is being said, this PR does not introduce new functionality to be tested. And, since probable luck of tests is not related to the Pull Request I completely against the idea. It makes sense when:
1. There is a good place for a TODO, like a non-implemented method
2. Strong relation with a Pull Request

That is why I mind it as absolutely inconvenient in this case

**Comment by amihaiemil** [comment]
@Sammers21 You don't understand. We are trying to enforce **discipline**. 

You work from an Office now. This team has been working together for years without an office, without a single meeting, or skype call, or chat. Just Github comments.\

You cannot achieve that without discipline. I think I've explained clearly why the TODO was needed.

**Comment by amihaiemil** [comment]
@Sammers21 That's why I insist on writing good description, making small PRs with unit tests included and add TODOs. 

So everything is easily traceable. New developers clearly see what has been done and why, what are the future tasks etc. Bugs are also much easier to track down and fix. 

You cannot achieve all this without proper discipline. Again, working for years in a distributed team around the World, with not a single meeting or Skype Call.

**Comment by 0crat** [comment]
Job was finished in 18 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Order was finished: +20 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by g4s8** [comment]
@amihaiemil merge permissions were removed from @Sammers21 now it should be ok
